### PR TITLE
[Local Catalog] Incremental sync: persist incrementally changed products/variations to storage

### PR DESCRIPTION
**Pull request by jaclync**
For WOOMOB-1298

- Based on https://github.com/woocommerce/woocommerce-ios/pull/16102
- ⚠️  Please review after https://github.com/woocommerce/woocommerce-ios/pull/16102 is approved

## Description

This PR implements persistence capabilities for the POS catalog incremental sync service. The changes enable the service to save incremental catalog updates (products and variations) to the database after fetching them from the API.

Key changes:
- Added `persistIncrementalCatalogData` method to `POSCatalogPersistenceService` that handles upsert operations for products and variations
- Updated `POSCatalogIncrementalSyncService` to integrate with the persistence service and save data after successful API fetches
- Made database column properties public in persisted model classes to support filtering operations
- Added test coverage for both incremental persistence functionality and error handling scenarios

The implementation ensures that:
- Products and variations are inserted or updated as needed
- Related attributes and images are properly replaced during updates
- Database operations are atomic and handle failures gracefully
- Sync dates are only updated after successful persistence to maintain data consistency

## Steps to reproduce

Just CI is sufficient, as the full/incremental services aren't integrated with the app yet.

## Testing information

I tested that incremental sync worked after a full sync for a big site, making changes to at least one product and variation while full sync is in progress:

```swift
Task {
   do {
        let date = Date()
        let catalog = try await syncService.startFullSync(for: siteID)
        print("Full sync completed with \(catalog.products.count) products and \(catalog.variations.count) variations for siteID \(siteID)")

        try await incrementalSyncService.startIncrementalSync(for: siteID, lastFullSyncDate: date)
        print("Incremental sync completed")
    } catch {
        print("Full sync failed for siteID \(siteID) with error: \(error)")
    }
}
```

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          2 Errors
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">PR requires at least one label.</td>
    </tr>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Modules/Sources/Yosemite/Tools/POS/POSCatalogPersistenceService.swift [hunk 1]
```diff
@@ -68,6 +74,70 @@ final class POSCatalogPersistenceService: POSCatalogPersistenceServiceProtocol {
...
+
+            try PersistedProductVariationImage
+                .filter(sql: "\(PersistedProductVariationImage.Columns.productVariationID.name) IN (\(variationIDs))")
+                .deleteAll(db)
+            for image in catalog.variationImagesToPersist {
+                try image.insert(db, onConflict: .replace)
+            }
+
+            try PersistedProductVariationAttribute
+                .filter(sql: "\(PersistedProductVariationAttribute.Columns.productVariationID.name) IN (\(variationIDs))")
```
Comment (jaclync) [inline Modules/Sources/Yosemite/Tools/POS/POSCatalogPersistenceService.swift]: This is an example of the last resort in p1757321561703349/1757321341.670339-slack-C070SJRA8DP. Please feel free to suggest a better way to do this. The corresponding foreign key column constant needs to be marked as public.

What I've tried:
- `.filter { variationIDs.contains($0.productVariationID) }`
- `.filter { variationIDs.contains(PersistedProductVariationAttribute.Columns.productVariationID) }`

The issue is that `$0.productVariationID` or `PersistedProductVariationAttribute.Columns.productVariationID` is of `Column` type, and only basic operators like `==` are supported. `.filter { $0.productVariationID == 2 }` works.

We could delete all existing attributes per product/variation in a prettier way, but deleting from `productVariationID in (variationIDs)` would be the most performant. WDYT? I personally think it's fine having raw SQL + unit tests if it helps with performance.

Comment (joshheald) [reply Modules/Sources/Yosemite/Tools/POS/POSCatalogPersistenceService.swift]: From what I've read and tried, it should be perfectly quick to do it in a loop, as long as that loop is inside the transaction, i.e.
```
           for variation in catalog.variations {
               try PersistedProductVariationAttribute
                    .filter { $0.productVariationID == variation.productVariationID }
                    .deleteAll(db)
            }
```

That would probably be ok for an incremental sync, which should be a relatively small change set. If it's not a small change, we might be better off doing a full sync anyway?

That said, there's absolutely nothing wrong with using SQL directly. We need to take care to correctly escape it (though type safety helps a lot when we're using Int64 IDs!)

Perhaps see whether you can detect any performance difference between the two?

File: Modules/Sources/Storage/GRDB/Model/PersistedProductVariationImage.swift [hunk 2]
```diff
@@ -34,7 +34,7 @@ extension PersistedProductVariationImage: FetchableRecord, PersistableRecord {
 
     public enum Columns {
         static let id = Column(CodingKeys.id)
-        static let productVariationID = Column(CodingKeys.productVariationID)
+        public static let productVariationID = Column(CodingKeys.productVariationID)
```
Comment (joshheald) [inline Modules/Sources/Storage/GRDB/Model/PersistedProductVariationImage.swift:37]: We can make all these public but I'll do it in a future PR. 

I'm sure that periphery would moan about it, but actually it's part of the public interface for the persisted classes. Since the error messages are unclear, it's a better experience to make them public ahead of use.

File: Modules/Sources/Yosemite/Tools/POS/POSCatalogPersistenceService.swift [hunk 3]
```diff
@@ -68,6 +74,70 @@ final class POSCatalogPersistenceService: POSCatalogPersistenceServiceProtocol {
...
+            try PersistedProductImage
+                .filter(sql: "\(PersistedProductImage.Columns.productID.name) IN (\(productIDs))")
+                .deleteAll(db)
+            for image in catalog.productImagesToPersist {
+                try image.insert(db, onConflict: .replace)
+            }
+
+            try PersistedProductAttribute
+                .filter(sql: "\(PersistedProductAttribute.Columns.productID.name) IN (\(productIDs))")
+                .deleteAll(db)
```
Comment (joshheald) [inline Modules/Sources/Yosemite/Tools/POS/POSCatalogPersistenceService.swift]: For deletions, the alternative approach is to do `PersistedProductImage.filter { $0.productID == product.productID }.deleteAll(db)` in the product loop above.

I think when we move to compound keys (including the site ID) this might be an easier approach than constructing SQL strings for compound keys.

File: Modules/Sources/Yosemite/Tools/POS/POSCatalogPersistenceService.swift [hunk 4]
```diff
@@ -68,6 +74,70 @@ final class POSCatalogPersistenceService: POSCatalogPersistenceServiceProtocol {
...
+                try variation.insert(db, onConflict: .replace)
+            }
+
+            let variationIDs = catalog.variations.map { $0.productVariationID }.map { String($0) }.joined(separator: ",")
+
+            try PersistedProductVariationImage
+                .filter(sql: "\(PersistedProductVariationImage.Columns.productVariationID.name) IN (\(variationIDs))")
+                .deleteAll(db)
+            for image in catalog.variationImagesToPersist {
+                try image.insert(db, onConflict: .replace)
```
Comment (joshheald) [inline Modules/Sources/Yosemite/Tools/POS/POSCatalogPersistenceService.swift:116]: For the full sync, I used `onConflict: ignore`. We should be consistent – it doesn't matter which we go with for images or attributes (when we support global attributes in future.) Either first wins, or last wins.

For variations on an incremental sync, it does matter because we'll definitely want to replace... so perhaps we should use replace everywhere?

File: Modules/Sources/Yosemite/Tools/POS/POSCatalogPersistenceService.swift [hunk 5]
```diff
@@ -68,6 +74,70 @@ final class POSCatalogPersistenceService: POSCatalogPersistenceServiceProtocol {
                       "\(variationImageCount) variation images, \(variationAttributeCount) variation attributes")
         }
     }
+
+    func persistIncrementalCatalogData(_ catalog: POSCatalog, siteID: Int64) async throws {
+        DDLogInfo("💾 Persisting incremental catalog with \(catalog.products.count) products and \(catalog.variations.count) variations")
+
+        try await grdbManager.databaseConnection.write { db in
+            let site = PersistedSite(id: siteID)
+            try site.insert(db, onConflict: .ignore)
```
Comment (joshheald) [inline Modules/Sources/Yosemite/Tools/POS/POSCatalogPersistenceService.swift]: Should we ever insert a Site for an incremental persist? Perhaps this won't ever do anything but is good to have for safety

File: Modules/Tests/YosemiteTests/Tools/POS/POSCatalogPersistenceServiceTests.swift [hunk 6]
```diff
@@ -221,4 +221,278 @@ struct POSCatalogPersistenceServiceTests {
...
+
+            let attributes = try PersistedProductImage.fetchAll(db).sorted(by: { $0.id < $1.id })
+            #expect(attributes[0].src == "https://example.com/image1-1.jpg")
+            #expect(attributes[0].productID == 1)
+            #expect(attributes[1].src == "https://example.com/image3.jpg")
+            #expect(attributes[1].productID == 1)
+        }
+    }
+
+    @Test func persistIncrementalCatalogData_upserts_existing_and_new_products() async throws {
```
Comment (joshheald) [inline Modules/Tests/YosemiteTests/Tools/POS/POSCatalogPersistenceServiceTests.swift]: Strictly, I don't think it does.

Upsert is `INSERT x ON CONFLICT DO UPDATE SET x.name = name {etc}` – there is an `upsert(db)` function we could use for this, at least for mutable persistable records.

Currently we use `insert(db, onConflict: .replace)`, which generates `INSERT x ON CONFLICT DO REPLACE`.

AIUI, the main difference is the hooks that are triggered. With replace, records are deleted and recreated, so the deletions will cascade. For upsert, they're just updated so none of the deletion hooks run.

**Review by joshheald — APPROVED** [review]
Some comments in line, but looks good, thanks!

File: Modules/Sources/Yosemite/Tools/POS/POSCatalogPersistenceService.swift (hunk 1 above)
Comment (jaclync) [reply Modules/Sources/Yosemite/Tools/POS/POSCatalogPersistenceService.swift]: > Perhaps see whether you can detect any performance difference between the two?

With the test case at the bottom, for 1000 products each with 3 attributes and 2000 variations each with 2 attributes to replace, the average runtime from 100 runs was:

- SQL version to delete all attributes in a set of IDs: 0.168967 seconds
- For loop to delete attributes for each product/variation ID: 0.284567 seconds

=> there is ~0.12s difference deleting 7k attributes from the test case, not a big difference.

> That would probably be ok for an incremental sync, which should be a relatively small change set. If it's not a small change, we might be better off doing a full sync anyway?

Yea, given that it's quite an edge case with a big data set from an incremental sync, the performance difference is negligible for most cases. I will update the implementation to delete existing image(s)/attributes in a for loop.

<details>
<summary>Test function</summary>
```swift
    @Test func persistIncrementalCatalogData_performance_with_large_dataset() async throws {
        // Given - create a large dataset with existing products and variations that will be replaced
        let existingProductCount = 1000
        let existingVariationCount = 2000
        
        // Create existing products with attributes
        var existingProducts: [POSProduct] = []
        for i in 1...existingProductCount {
            let attributes = [
                ProductAttribute.fake().copy(name: "Color", options: ["Red", "Blue", "Green"]),
                ProductAttribute.fake().copy(name: "Size", options: ["S", "M", "L", "XL"]),
                ProductAttribute.fake().copy(name: "Material", options: ["Cotton", "Polyester"])
            ]
            let product = POSProduct.fake().copy(
                siteID: sampleSiteID,
                productID: Int64(i),
                name: "Product \(i)",
                attributes: attributes
            )
            existingProducts.append(product)
        }
        
        // Create existing variations with attributes
        var existingVariations: [POSProductVariation] = []
        for i in 1...existingVariationCount {
            let productID = Int64((i % existingProductCount) + 1)
            let attributes = [
                ProductVariationAttribute.fake().copy(name: "Color", option: "Red"),
                ProductVariationAttribute.fake().copy(name: "Size", option: "M")
            ]
            let variation = POSProductVariation.fake().copy(
                siteID: sampleSiteID,
                productID: productID,
                productVariationID: Int64(i),
                attributes: attributes,
                price: "\(10 + i).00"
            )
            existingVariations.append(variation)
        }
        
        // Insert existing data
        let existingCatalog = POSCatalog(products: existingProducts, variations: existingVariations)
        try await sut.replaceAllCatalogData(existingCatalog, siteID: sampleSiteID)
        
        // Create update catalog with modified data
        var updatedProducts: [POSProduct] = []
        for i in 1...existingProductCount {
            let attributes = [
                ProductAttribute.fake().copy(name: "Color", options: ["Purple", "Orange", "Yellow"]),
                ProductAttribute.fake().copy(name: "Size", options: ["XS", "S", "M", "L", "XL", "XXL"]),
                ProductAttribute.fake().copy(name: "Style", options: ["Casual", "Formal"])
            ]
            let product = POSProduct.fake().copy(
                siteID: sampleSiteID,
                productID: Int64(i),
                name: "Updated Product \(i)",
                attributes: attributes
            )
            updatedProducts.append(product)
        }
        
        var updatedVariations: [POSProductVariation] = []
        for i in 1...existingVariationCount {
            let productID = Int64((i % existingProductCount) + 1)
            let attributes = [
                ProductVariationAttribute.fake().copy(name: "Color", option: "Purple"),
                ProductVariationAttribute.fake().copy(name: "Size", option: "L"),
                ProductVariationAttribute.fake().copy(name: "Style", option: "Casual")
            ]
            let variation = POSProductVariation.fake().copy(
                siteID: sampleSiteID,
                productID: productID,
                productVariationID: Int64(i),
                attributes: attributes,
                price: "\(20 + i).00"
            )
            updatedVariations.append(variation)
        }
        
        let updateCatalog = POSCatalog(products: updatedProducts, variations: updatedVariations)
        
        // When - measure performance of incremental update
        let startTime = Date()
        try await sut.persistIncrementalCatalogData(updateCatalog, siteID: sampleSiteID)
        let endTime = Date()
        let timeElapsed = endTime.timeIntervalSince(startTime)
        
        // Then - verify the operation completed in reasonable time and data is correct
        print("persistIncrementalCatalogData with \(existingProductCount) products and \(existingVariationCount) variations took \(timeElapsed) seconds")
        
        try await grdbManager.databaseConnection.read { db in
            let productCount = try PersistedProduct.fetchCount(db)
            let variationCount = try PersistedProductVariation.fetchCount(db)
            let attributeCount = try PersistedProductAttribute.fetchCount(db)
            let variationAttributeCount = try PersistedProductVariationAttribute.fetchCount(db)
            
            #expect(productCount == existingProductCount)
            #expect(variationCount == existingVariationCount)
            #expect(attributeCount == existingProductCount * 3) // 3 attributes per product
            #expect(variationAttributeCount == existingVariationCount * 3) // 3 attributes per variation
            
            // Verify data was actually updated by checking a sample
            let sampleProduct = try PersistedProduct.filter(sql: "\(PersistedProduct.Columns.id.name) = 1").fetchOne(db)
            #expect(sampleProduct?.name == "Updated Product 1")
            
            let sampleVariation = try PersistedProductVariation.filter(sql: "\(PersistedProductVariation.Columns.id.name) = 1").fetchOne(db)
            #expect(sampleVariation?.price == "21.00")
            
            // Verify attributes were replaced correctly
            let productAttributes = try PersistedProductAttribute.filter(sql: "\(PersistedProductAttribute.Columns.productID.name) = 1").fetchAll(db)
            let attributeNames = Set(productAttributes.map { $0.name })
            #expect(attributeNames == Set(["Color", "Size", "Style"]))
        }
        
        // Performance assertion - should complete within 10 seconds for this dataset size
        #expect(timeElapsed < 10.0, "Performance test failed: operation took \(timeElapsed) seconds, expected < 10 seconds")
    }
```
</details>

File: Modules/Sources/Yosemite/Tools/POS/POSCatalogPersistenceService.swift (hunk 5 above)
Comment (jaclync) [reply Modules/Sources/Yosemite/Tools/POS/POSCatalogPersistenceService.swift]: Good catch - no, it's not necessary to insert a Site in an incremental sync as a full sync is a prerequisite. I removed it and updated test cases that depend on the Site insertion in https://github.com/woocommerce/woocommerce-ios/pull/16106/commits/cdd580d7cb758ae941e85719a70ff44e2b2c1c4c.

File: Modules/Tests/YosemiteTests/Tools/POS/POSCatalogPersistenceServiceTests.swift (hunk 6 above)
Comment (jaclync) [reply Modules/Tests/YosemiteTests/Tools/POS/POSCatalogPersistenceServiceTests.swift]: Good to know about the difference. I renamed the products/variations test cases to mention "replace" instead of "upsert" in https://github.com/woocommerce/woocommerce-ios/pull/16106/commits/7b371910f91026ef6044503bbf90f580773bf10a and https://github.com/woocommerce/woocommerce-ios/pull/16106/commits/20b59103986da832531d77e0a66926d784e365ed.

File: Modules/Sources/Yosemite/Tools/POS/POSCatalogPersistenceService.swift (hunk 4 above)
Comment (jaclync) [reply Modules/Sources/Yosemite/Tools/POS/POSCatalogPersistenceService.swift:116]: Yea, incremental sync requires `relace` on conflict. For full sync, like you said, either works as we delete all the data at the beginning. It only matters if there are duplicate products/variations in the catalog (which is not impossible from the "Large fun testing" test site) with different values, but there's no way to know whether the first or last is correct in this edge case.

I replaced the `ignore` with `replace` on conflict strategy in full sync in
https://github.com/woocommerce/woocommerce-ios/pull/16106/commits/7162e20a2aeaaa912b06ded5c1f8912c9357a9a2.

File: Modules/Sources/Yosemite/Tools/POS/POSCatalogPersistenceService.swift (hunk 1 above)
Comment (jaclync) [reply Modules/Sources/Yosemite/Tools/POS/POSCatalogPersistenceService.swift]: Updated filtering by SQL with for loops comparing with ID in https://github.com/woocommerce/woocommerce-ios/pull/16106/commits/7dbd5b3a3c419d24253f2f5e5e082194c29a90b6.

File: Modules/Sources/Yosemite/Tools/POS/POSCatalogPersistenceService.swift (hunk 3 above)
Comment (jaclync) [reply Modules/Sources/Yosemite/Tools/POS/POSCatalogPersistenceService.swift]: Similar to above, I updated filtering by SQL with for loops comparing with ID in https://github.com/woocommerce/woocommerce-ios/pull/16106/commits/7dbd5b3a3c419d24253f2f5e5e082194c29a90b6.

**Comment by jaclync** [comment]
Thanks for the review and GRDB insights! I believe I responded to all the comments, merging now but feel free to leave further suggestions I can make changes separately.
